public/scripts/nai-settings.js

```diff
@@ -624,11 +630,67 @@ export async function generateNovelWithStreaming(generate_data, signal) {
                 text += data.token;
             }
 
-            yield { text, swipes: [] };
+            yield { text, swipes: [], logprobs: parseNovelAILogprobs(data.logprobs) };
         }
     };
 }
 
+/**
+ * A single token's ID.
+ * @typedef {[number]} TokenIdEntry
+ */
+/**
+ * A single token's log probabilities. The first element is before repetition
+ * penalties and samplers are applied, the second is after.
+ * @typedef {[number, number]} LogprobsEntry
+ */
+/**
+ * Combination of token ID and its corresponding log probabilities.
+ * @typedef {[TokenIdEntry, LogprobsEntry]} TokenLogprobTuple
+ */
+/**
+ * Represents all logprob data for a single token, including its
+ * before, after, and the ultimately selected token.
+ * @typedef {Object} NAITokenLogprobs
+ * @property {TokenLogprobTuple[]} chosen - always length 1
+ * @property {TokenLogprobTuple[]} before - always however many logprobs were requested
+ * @property {TokenLogprobTuple[]} after - may be fewer than requested if some tokens were eliminated
+ */
+/**
+ * parseNovelAILogprobs converts NovelAI's logprobs format into the format we use
+ * in the Alternative Tokens view.
+ * @param {NAITokenLogprobs} data - NovelAI logprobs for one token
+ * @returns {import('logprobs.js').TokenLogprobs | null} converted logprobs for one token
+ */
+export function parseNovelAILogprobs(data) {
+    if (!data) {
+        return null;
+    }
+    const befores = data.before.map(([[tokenId], [before, _]]) => [tokenId, before]);
+    const afters = data.after.map(([[tokenId], [_, after]]) => [tokenId, after]);
```

Comment:
The NAI logprobs data structure is nested arrays and tuples all the way down and was driving me nuts, so I exhaustively typed it to hopefully make it less confusing for someone in the future.